src/GraphQL/Internal/Validation.hs

```diff
@@ -386,12 +472,77 @@ resolveFragmentDefinitions allFragments =
           modify (Set.insert name)
           FragmentSpread name directives <$> resolveFragment' definition
 
+-- * Variables
+
+data VariableDefinition = VariableDefinition Variable AST.Type (Maybe Value) deriving (Eq, Ord, Show)
+
+type VariableDefinitions = Map Variable VariableDefinition
+
+getDefinedVariables :: VariableDefinitions -> Set Variable
+getDefinedVariables = Map.keysSet
+
+-- | A GraphQL value which might contain some defined variables.
+type VariableValue = Value' (Either VariableDefinition ConstScalar)
+
+emptyVariableDefinitions :: VariableDefinitions
+emptyVariableDefinitions = mempty
+
+-- | Ensure that a set of variable definitions is valid.
+validateVariableDefinitions :: [AST.VariableDefinition] -> Validation VariableDefinitions
+validateVariableDefinitions vars = do
+  validatedDefns <- traverse validateVariableDefinition vars
+  let items = [ (name, defn) | defn@(VariableDefinition name _ _) <- validatedDefns]
+  mapErrors DuplicateVariableDefinition (makeMap items)
+
+-- | Ensure that a variable definition is a valid one.
+validateVariableDefinition :: AST.VariableDefinition -> Validation VariableDefinition
+validateVariableDefinition (AST.VariableDefinition name varType value) =
+  VariableDefinition name varType <$> (traverse validateDefaultValue value)
+
+-- | Ensure that a default value contains no variables.
+validateDefaultValue :: AST.DefaultValue -> Validation Value
+validateDefaultValue defaultValue =
+  case astToVariableValue defaultValue of
+    Nothing -> throwE $ InvalidValue defaultValue
+    Just value ->
+      for value $ \scalar -> case scalar of
```

Comment:
Is this a normal use for `for`? Reads a bit weird to traverse a single-element container...

`\scalar -> .. case scalar of ` might be a case for `LambdaCase` if that's your thing.